bundles/remote_services/discovery_zeroconf/src/discovery_zeroconf_announcer.c

```diff
@@ -0,0 +1,533 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <discovery_zeroconf_announcer.h>
+#include <discovery_zeroconf_constants.h>
+#include <dns_sd.h>
+#include <endpoint_listener.h>
+#include <remote_constants.h>
+#include <celix_utils.h>
+#include <celix_properties.h>
+#include <celix_constants.h>
+#include <celix_threads.h>
+#include <celix_bundle_context.h>
+#include <celix_string_hash_map.h>
+#include <celix_array_list.h>
+#include <celix_log_helper.h>
+#include <celix_types.h>
+#include <celix_errno.h>
+#include <celix_build_assert.h>
+#include <netinet/in.h>
+#include <net/if.h>
+#include <sys/eventfd.h>
+#include <sys/select.h>
+#include <sys/param.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+
+
+#define DZC_MAX_CONFLICT_CNT 256
+
+//According to rfc6763, Using TXT records larger than 1300 bytes is NOT RECOMMENDED
+#define DZC_MAX_TXT_RECORD_SIZE 1300
+
+struct discovery_zeroconf_announcer {
+    celix_bundle_context_t *ctx;
+    celix_log_helper_t *logHelper;
+    char fwUuid[64];
+    endpoint_listener_t epListener;
+    long epListenerSvcId;
+    DNSServiceRef sharedRef;
+    int eventFd;
+    celix_thread_t refreshEPThread;
+    celix_thread_mutex_t mutex;//projects below
+    bool running;
+    celix_string_hash_map_t *endpoints;//key:endpoint id, val:announce_endpoint_entry_t*
+    celix_array_list_t *revokedEndpoints;
+};
+
+typedef struct announce_endpoint_entry {
+    celix_properties_t *properties;
+    DNSServiceRef registerRef;
+    unsigned int uid;
+    int ifIndex;
+    const char *serviceName;
+    char serviceType[64];
+    bool announced;
+}announce_endpoint_entry_t;
+
+
+static void discoveryZeroconfAnnouncer_eventNotify(discovery_zeroconf_announcer_t *announcer);
+static  celix_status_t discoveryZeroconfAnnouncer_endpointAdded(void *handle, endpoint_description_t *endpoint, char *matchedFilter);
+static celix_status_t discoveryZeroconfAnnouncer_endpointRemoved(void *handle, endpoint_description_t *endpoint, char *matchedFilter);
+static void *discoveryZeroconfAnnouncer_refreshEndpointThread(void *data);
+
+
+celix_status_t discoveryZeroconfAnnouncer_create(celix_bundle_context_t *ctx, celix_log_helper_t *logHelper, discovery_zeroconf_announcer_t **announcerOut) {
+    celix_status_t status = CELIX_SUCCESS;
+    discovery_zeroconf_announcer_t *announcer = calloc(1, sizeof(*announcer));
+    assert(announcer != NULL);
+    announcer->ctx = ctx;
+    announcer->logHelper = logHelper;
+    announcer->sharedRef = NULL;
+
+    announcer->eventFd = eventfd(0, 0);
```

Comment:
I am not really familiar with eventfd, but this seems nice.

Could this also replace the pipe usage in shell_tui? 

In shell_tui a pipe used to provide a fd to stop the text ui thread:
https://github.com/apache/celix/blob/master/bundles/shell/shell_tui/src/shell_tui.c#L121-L125
https://github.com/apache/celix/blob/master/bundles/shell/shell_tui/src/shell_tui.c#L141